api_tests/base/test_auth.py

```diff
@@ -3,271 +3,300 @@
-
-    def test_missing_credential_fails(self):
-        res = self.app.get(self.unreachable_url, auth=None, expect_errors=True)
-        assert_equal(res.status_code, 401)
-        assert_equal(res.json.get("errors")[0]['detail'], 'Authentication credentials were not provided.')
-
-    def test_invalid_credential_fails(self):
-        res = self.app.get(self.unreachable_url, auth=(self.user1.username, 'invalid password'), expect_errors=True)
-        assert_equal(res.status_code, 401)
-        assert_equal(res.json.get("errors")[0]['detail'], 'Invalid username/password.')
-
-    def test_valid_credential_authenticates_and_has_permissions(self):
-        res = self.app.get(self.reachable_url, auth=self.user1.auth)
-        assert_equal(res.status_code, 200, msg=res.json)
-
-    def test_valid_credential_authenticates_but_user_lacks_object_permissions(self):
-        res = self.app.get(self.unreachable_url, auth=self.user1.auth, expect_errors=True)
-        assert_equal(res.status_code, 403, msg=res.json)
-
-    def test_valid_credential_but_twofactor_required(self):
-        user1_addon = self.user1.get_or_add_addon('twofactor')
-        user1_addon.totp_drift = 1
-        user1_addon.totp_secret = self.TOTP_SECRET
-        user1_addon.is_confirmed = True
-        user1_addon.save()
-
-        res = self.app.get(self.reachable_url, auth=self.user1.auth, expect_errors=True)
-        assert_equal(res.status_code, 401)
-        assert_equal(res.headers['X-OSF-OTP'], 'required; app')
-        assert_equal(res.json.get("errors")[0]['detail'], 'Must specify two-factor authentication OTP code.')
-
-    def test_valid_credential_twofactor_invalid_otp(self):
-        user1_addon = self.user1.get_or_add_addon('twofactor')
-        user1_addon.totp_drift = 1
-        user1_addon.totp_secret = self.TOTP_SECRET
-        user1_addon.is_confirmed = True
-        user1_addon.save()
-
-        res = self.app.get(self.reachable_url, auth=self.user1.auth, headers={'X-OSF-OTP': 'invalid otp'}, expect_errors=True)
-        assert_equal(res.status_code, 401)
-        assert_true('X-OSF-OTP' not in res.headers)
-        assert_equal(res.json.get("errors")[0]['detail'], 'Invalid two-factor authentication OTP code.')
-
-    def test_valid_credential_twofactor_valid_otp(self):
-        user1_addon = self.user1.get_or_add_addon('twofactor')
-        user1_addon.totp_drift = 1
-        user1_addon.totp_secret = self.TOTP_SECRET
-        user1_addon.is_confirmed = True
-        user1_addon.save()
-
-        res = self.app.get(self.reachable_url, auth=self.user1.auth, headers={'X-OSF-OTP': _valid_code(self.TOTP_SECRET)})
-        assert_equal(res.status_code, 200)
-
-
-class TestOAuthValidation(ApiTestCase):
+    # Test projects for which a given user DOES and DOES NOT  have appropriate permissions
+    @pytest.fixture()
+    def project_reachable(self, user_one):
+        return ProjectFactory(title="Private Project User 1", is_public=False, creator=user_one)
```

Comment:
Single quotes. They show up in a number of places, might even consider running autopep8 on the file

Reply:
autopep8 is awesome, magic. Just ran it on all the test files. Thanks